### PR TITLE
replicaset: add command 'roles remove'

### PR DESCRIPTION
**Pull request by themilchenko**
@TarantoolBot document
Title: `tt replicaset roles remove` removes roles in the tarantool replicaset
with cluster config (3.0) or cartridge orchestrator.

This patch introduces new command for the replicaset module.

```
$ tt rs roles remove [--cartridge|--config|--custom] [-f] [--timeout secs]
    <APP_NAME:INSTANCE_NAME> <ROLE_NAME> [flags]
```

It is possible to provide `cartridge`, `config` or `custom` flag to explicitly state which orchestrator to use. ROLE_NAME is a role to remove from local cluster config in case of `cluster config` orchestrator and directly from all instances of replicaset in case of `cartridge` orchestrator. Command supports `cartridge` and `cluster config` orchestrators only for the entire application. INSTANCE_NAME works only for `cluster config` to provide instance  ame to remove role from.

There are flags supported by this command:
- `--global (-G)` for a global scope to add a role (only for `cluster config`
  orchestrator);
- `--instance (-i) string` for an application name target to specify
  an instance to add a role;
- `--replicaset (-r) string` for an application name target to specify
  a replicaset to add a role;
- `--group (-g) string` for an application name target to specify a group
  to specify a group to add a role (only for `cluster config`);
- `--force (-f)` skips instances not found locally in `cluster
  config` orchestrator.

Closes https://github.com/tarantool/tt/issues/916

### CODE REVIEW

**Comment by oleg-jukovec** [comment]
Please, rebase to the `master` branch.

**Comment by themilchenko** [comment]
> Please, rebase to the `master` branch.

Rebased.

**Comment by oleg-jukovec** [comment]
Oh, three tests are flak.

File: cli/cmd/replicaset.go [hunk 1]
```diff
@@ -293,6 +294,44 @@ func newRolesAddCmd() *cobra.Command {
 	return cmd
 }
 
+// newRolesAddCmd creates a "replicaset roles add" command.
+func newRolesRemoveCmd() *cobra.Command {
```
Comment (oleg-jukovec) [inline cli/cmd/replicaset.go:298]:
```suggestion
// newRolesRemoveCmd creates a "replicaset roles remove" command.
func newRolesRemoveCmd() *cobra.Command {
```

File: cli/cmd/replicaset.go [hunk 2]
```diff
@@ -293,6 +294,44 @@ func newRolesAddCmd() *cobra.Command {
...
+		"global config context")
+
+	addOrchestratorFlags(cmd)
+	addTarantoolConnectFlags(cmd)
+	cmd.Flags().BoolVarP(&replicasetForce, "force", "f", false,
+		"to force a promotion:\n"+
+			"  * config: skip instances not found locally\n"+
+			"  * cartridge: force inconsistency")
+	cmd.Flags().IntVarP(&replicasetTimeout, "timeout", "",
+		replicasetcmd.DefaultTimeout, "adding timeout")
```
Comment (oleg-jukovec) [inline cli/cmd/replicaset.go:329]:
```suggestion
		replicasetcmd.DefaultTimeout, "removing timeout")
```

**Review by oleg-jukovec — COMMENTED** [review]
Please, add an entry to CHANGELOG.md

File: cli/replicaset/cartridge.go [hunk 3]
```diff
@@ -857,7 +872,7 @@ type cartridgeJoinServersOpts struct {
 type cartridgeEditReplicasetsOpts struct {
 	UUID             *string                    `msgpack:"uuid,omitempty"`
 	Alias            *string                    `msgpack:"alias,omitempty"`
-	Roles            []string                   `msgpack:"roles,omitempty"`
+	Roles            []string                   `msgpack:"roles"`
```
Comment (oleg-jukovec) [inline cli/replicaset/cartridge.go:875]: Why the change is required?

File: cli/replicaset/cmd/roles.go [hunk 4]
```diff
@@ -79,37 +79,48 @@ func RolesAdd(ctx RolesAddCtx) error {
 	statusReplicasets(replicasets)
 	fmt.Println()
 
+	action := []string{"Add", "to"}
+	if replicaset.IsChangeRolesFuncEqual(changeRoleFunc, replicaset.RemoveRole) {
+		action = []string{"Remove", "from"}
+	}
```
Comment (oleg-jukovec) [inline cli/replicaset/cmd/roles.go:85]: Comparing functions is usually a bad idea. At least, you should to do it via a switch with all cases.

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (themilchenko) [reply cli/replicaset/cartridge.go:875]: In case we delete last role we send empty []Roles and `omitempty` ignores it. As a result we cannot update roles.

File: cli/cmd/replicaset.go (hunk 2 above)
Comment (themilchenko) [reply cli/cmd/replicaset.go:329]: It is timeout for Lua eval request.

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (oleg-jukovec) [reply cli/replicaset/cartridge.go:875]: Won't roles be removed for other command that call the eval then?

File: cli/cmd/replicaset.go (hunk 1 above)
Comment (themilchenko) [reply cli/cmd/replicaset.go]: Corrected.

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (themilchenko) [reply cli/replicaset/cartridge.go:876]: Sorry, I wasn't right. Backed `omitempty` as it was and everything works in expected way.

File: cli/replicaset/cmd/roles.go (hunk 4 above)
Comment (themilchenko) [reply cli/replicaset/cmd/roles.go]: Refactored using polymorphism ideology.

File: cli/replicaset/roles.go [hunk 5]
```diff
@@ -5,18 +5,57 @@ import (
...
+)
+
+// ChangeRoleFunc is a function for adding/removing roles.
 type ChangeRoleFunc func([]string, string) ([]string, error)
 
-// AddRole is a function that implements addition of role.
+// RolesChangerAction implements changing (add or remove) of roles.
+type RolesChangerAction interface {
+	// ChangeRoleFunc is a function type for addition or removing a role.
+	ChangeRoleFunc([]string, string) ([]string, error)
```
Comment (oleg-jukovec) [inline cli/replicaset/roles.go:23]: It's just shorter and we don't lose anything.
```suggestion
	Change([]string, string) ([]string, error)
```

File: cli/replicaset/roles.go [hunk 6]
```diff
@@ -48,24 +87,32 @@ type RolesChangeCtx struct {
...
 }
 
-// RolesAdder is an interface for adding roles to a replicaset.
-type RolesAdder interface {
-	// RolesAdd adds role to a replicasets by its name.
-	RolesAdd(ctx RolesChangeCtx) error
+// RolesChanger is an interface for adding/removing roles to a replicaset.
+type RolesChanger interface {
+	// RolesChange adds/remove role to a replicasets by its name.
+	RolesChange(ctx RolesChangeCtx, changeRoleAction RolesChangerAction) error
```
Comment (oleg-jukovec) [inline cli/replicaset/roles.go:93]: The same.
```suggestion
	RolesChange(ctx RolesChangeCtx, action RolesChangerAction) error
```
Please, apply the change to the same code in the pull request.

File: cli/replicaset/roles.go [hunk 7]
```diff
@@ -5,18 +5,57 @@ import (
...
+// RoleAction is a type that describes an action that will
+// happen with Roles. It can be "add" or "remove".
+type RoleAction uint
+
+const (
+	AddAction RoleAction = iota
+	RemoveAction
+)
+
+// ChangeRoleFunc is a function for adding/removing roles.
```
Comment (oleg-jukovec) [inline cli/replicaset/roles.go:17]: Why do we need a function if we have an interface?

File: cli/replicaset/roles.go [hunk 8]
```diff
@@ -5,18 +5,57 @@ import (
...
 type ChangeRoleFunc func([]string, string) ([]string, error)
 
-// AddRole is a function that implements addition of role.
+// RolesChangerAction implements changing (add or remove) of roles.
+type RolesChangerAction interface {
```
Comment (oleg-jukovec) [inline cli/replicaset/roles.go:21]:
```suggestion
// RolesChanger implements changing (add or remove) of roles.
type RolesChanger interface {
```

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (themilchenko) [reply cli/replicaset/cartridge.go:876]: No, it won't. If we do not pass `roles` field it will be initialized with a nil value which is `nil`. So it won't erase existing roles.

File: cli/replicaset/roles.go (hunk 7 above)
Comment (themilchenko) [reply cli/replicaset/roles.go:17]: To name interface's method in function argument to not pass a whole set of methods. For example: 

https://github.com/tarantool/tt/blob/29379627ec9c41c63ce3d99e6791e6bd23ba20a8/cli/replicaset/cconfig.go#L755-L756

File: cli/replicaset/roles.go (hunk 8 above)
Comment (themilchenko) [reply cli/replicaset/roles.go:21]: We already have an interface with the same name.

File: cli/replicaset/roles.go (hunk 5 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Renamed.

File: cli/replicaset/roles.go (hunk 6 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Renamed.

File: cli/replicaset/roles.go (hunk 7 above)
Comment (oleg-jukovec) [reply cli/replicaset/roles.go:17]: Interfaces should not contain many methods, so I don't see any problem. Adding a function type for this is unnecessary.

**Review by oleg-jukovec — COMMENTED** [review]
Please, add comments to a new public API: methods/types/etc.

File: cli/replicaset/roles.go (hunk 8 above)
Comment (oleg-jukovec) [reply cli/replicaset/roles.go:21]: It looks like we have different levels of abstraction mixed in one place. It might be confusing, but let it be.

File: cli/replicaset/roles.go [hunk 9]
```diff
@@ -5,18 +5,57 @@ import (
...
+// RolesRemover is a struct that implements removing of role.
+type RolesRemover struct{}
+
+func (RolesRemover) Change(roles []string, r string) ([]string, error) {
+	return RemoveRole(roles, r)
+}
+
+func (RolesRemover) Action() RoleAction {
+	return RemoveAction
+}
```
Comment (oleg-jukovec) [inline cli/replicaset/roles.go:48]: It seems like you could add unit-tests for the code. And for AddRole/RemoveRole (do we need it in the public API?) functions too.

File: cli/replicaset/roles.go (hunk 7 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Removed function declaration and change arg on interface in related functions.

File: cli/replicaset/roles.go (hunk 8 above)
Comment (themilchenko) [reply cli/replicaset/roles.go:18]: Ok, then I'll just leave it at that.

File: cli/replicaset/roles.go (hunk 9 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Added tests and deleted exportable `AddRole`, `RemoveRole` functions.

File: cli/cmd/replicaset.go [hunk 10]
```diff
@@ -293,6 +294,44 @@ func newRolesAddCmd() *cobra.Command {
 	return cmd
 }
 
+// newRolesRemoveCmd creates a "replicaset roles remove" command.
+func newRolesRemoveCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use: "remove [--cartridge|--config|--custom] [-f] [--timeout secs]" +
+			"<APP_NAME:INSTANCE_NAME> <ROLE_NAME> [flags]",
+		Short: "Removes a role for the Cartridge and the Tarantool 3 orchestrator",
+		Long:  "Removes a role for the Cartridge and the Tarantool 3 orchestrator",
```
Comment (DerekBum) [inline cli/cmd/replicaset.go:303]:
```suggestion
		Short: "Removes a role for Cartridge and Tarantool 3 orchestrator",
		Long:  "Removes a role for Cartridge and Tarantool 3 orchestrator",
```

File: cli/cmd/replicaset.go [hunk 11]
```diff
@@ -402,8 +441,8 @@ func replicasetFillCtx(cmdCtx *cmdcontext.CmdCtx, ctx *replicasetCtx, args []str
 				}
 			}
 		}
-		// In case of adding a role when user may not provide an instance.
-		if cmdCtx.CommandName == "add" && ctx.InstName == "" {
+		// In case of adding/removing role when user may to not provide instance.
```
Comment (DerekBum) [inline cli/cmd/replicaset.go:444]:
```suggestion
		// In case of adding/removing role when user may not provide an instance.
```

File: cli/replicaset/cmd/roles.go [hunk 12]
```diff
@@ -42,8 +42,8 @@ type RolesAddCtx struct {
 	Timeout int
 }
 
-// RolesAdd adds role with provided path target to config.
-func RolesAdd(ctx RolesAddCtx) error {
+// RolesChange adds role with provided path target to config.
```
Comment (DerekBum) [inline cli/replicaset/cmd/roles.go:45]:
```suggestion
// RolesChange adds/removes role with provided path target to config.
```

File: cli/replicaset/configsource.go [hunk 13]
```diff
@@ -214,8 +214,8 @@ func (c *CConfigSource) Expel(ctx ExpelCtx) error {
 }
 
 // ChangeRole patches a config to add role to a config.
-func (c *CConfigSource) ChangeRole(ctx RolesChangeCtx, changeRoleFunc ChangeRoleFunc) error {
-	return c.patchConfigWithRoles(ctx, getCConfigRolesPath, changeRoleFunc, patchCConfigEditRole)
+func (c *CConfigSource) ChangeRole(ctx RolesChangeCtx, action RolesChangerAction) error {
```
Comment (DerekBum) [inline cli/replicaset/configsource.go:217]: Comment for this function might also need an update:
`ChangeRole patches a config to add/remove role.`

File: cli/replicaset/roles.go [hunk 14]
```diff
@@ -5,19 +5,44 @@ import (
 	"slices"
 )
 
-// ChangeRoleFunc is a function type for addition or removing a role.
-type ChangeRoleFunc func([]string, string) ([]string, error)
+// RoleAction is a type that describes an action that will
+// happen with Roles. It can be "add" or "remove".
+type RoleAction uint
```
Comment (DerekBum) [inline cli/replicaset/roles.go:10]: `RoleAction` has `uint` type, but in the doc above you wrote that `It can be "add" or "remove"` (string values). Can you specify which `RoleAction` value (0/1) corresponds to which action (or omit mentioning them)?

File: cli/replicaset/roles.go [hunk 15]
```diff
@@ -48,24 +78,32 @@ type RolesChangeCtx struct {
 	Timeout int
 }
 
-// RolesAdder is an interface for adding roles to a replicaset.
-type RolesAdder interface {
-	// RolesAdd adds role to a replicasets by its name.
-	RolesAdd(ctx RolesChangeCtx) error
+// RolesChanger is an interface for adding/removing roles to a replicaset.
+type RolesChanger interface {
+	// RolesChange adds/remove role to a replicasets by its name.
```
Comment (DerekBum) [inline cli/replicaset/roles.go:83]:
```suggestion
	// RolesChange adds/removes role for a replicasets by its name.
```

File: cli/replicaset/roles.go [hunk 16]
```diff
@@ -48,24 +78,32 @@ type RolesChangeCtx struct {
 	Timeout int
 }
 
-// RolesAdder is an interface for adding roles to a replicaset.
-type RolesAdder interface {
-	// RolesAdd adds role to a replicasets by its name.
-	RolesAdd(ctx RolesChangeCtx) error
+// RolesChanger is an interface for adding/removing roles to a replicaset.
```
Comment (DerekBum) [inline cli/replicaset/roles.go:81]:
```suggestion
// RolesChanger is an interface for adding/removing roles for a replicaset.
```

File: test/integration/replicaset/test_replicaset_roles_add.py [hunk 17]
```diff
@@ -239,7 +238,7 @@ def test_replicaset_cconfig_roles_add(
         role_name="failover-coordinator",
         inst_flg="s1-master",
         err_msg=("cannot pass the instance or --instance (-i) flag due to cluster"
-                 " with cartridge orchestrator can't add role into instance scope")
+                 " with cartridge orchestrator can't add/remove role into instance scope")
```
Comment (DerekBum) [inline test/integration/replicaset/test_replicaset_roles_add.py:241]:
```suggestion
                 " with cartridge orchestrator can't add/remove role for instance scope")
```

File: cli/replicaset/cartridge.go [hunk 18]
```diff
@@ -736,7 +731,28 @@ func (c *CartridgeApplication) RolesAdd(ctx RolesChangeCtx) error {
 	return nil
 }
 
-// getReplicasetByAlias searches for a replicaset by its alias in discovered slice
+// getRolesChangedOpts adds/removes role from replicaset roles list and returns
+// options for updating a replciaset.
```
Comment (DerekBum) [inline cli/replicaset/cartridge.go:735]:
```suggestion
// getRolesChangedOpts adds/removes role for replicaset roles list and returns
// options for updating a replciaset.
```

File: cli/replicaset/cmd/roles.go [hunk 19]
```diff
@@ -79,37 +79,48 @@ func RolesAdd(ctx RolesAddCtx) error {
...
+			changeRoleAction.Action() == replicaset.RemoveAction {
+			return fmt.Errorf("cannot provide vshard-group by removing role")
+		}
+		log.Infof("%s role %s %s group: %s", action[0], ctx.RoleName, action[1], ctx.GroupName)
 	}
 	if ctx.InstName != "" {
 		if orchestratorType == replicaset.OrchestratorCartridge {
 			return fmt.Errorf("cannot pass the instance or --instance (-i) flag due to cluster" +
-				" with cartridge orchestrator can't add role into instance scope")
+				" with cartridge orchestrator can't add/remove role into instance scope")
```
Comment (DerekBum) [inline cli/replicaset/cmd/roles.go:104]:
```suggestion
				" with cartridge orchestrator can't add/remove role for instance scope")
```

File: cli/replicaset/cmd/roles.go [hunk 20]
```diff
@@ -10,9 +10,9 @@ import (
 	libcluster "github.com/tarantool/tt/lib/cluster"
 )
 
-// RolesAddCtx describes the context to add a role to
+// RolesChangeCtx describes the context to add/remove role to
```
Comment (DerekBum) [inline cli/replicaset/cmd/roles.go:13]:
```suggestion
// RolesChangeCtx describes the context to add/remove role for
```

File: cli/cmd/replicaset.go (hunk 10 above)
Comment (themilchenko) [reply cli/cmd/replicaset.go]: Deleted.

File: cli/cmd/replicaset.go (hunk 11 above)
Comment (themilchenko) [reply cli/cmd/replicaset.go]: Added.

File: cli/replicaset/cmd/roles.go (hunk 12 above)
Comment (themilchenko) [reply cli/replicaset/cmd/roles.go]: Added.

File: cli/replicaset/configsource.go (hunk 13 above)
Comment (themilchenko) [reply cli/replicaset/configsource.go:217]: Yes, corrected, thank you!

File: cli/replicaset/roles.go (hunk 14 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Added explanation for each action.

File: cli/replicaset/roles.go (hunk 15 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Corrected.

File: cli/replicaset/roles.go (hunk 16 above)
Comment (themilchenko) [reply cli/replicaset/roles.go]: Edited.

File: cli/replicaset/cartridge.go (hunk 18 above)
Comment (themilchenko) [reply cli/replicaset/cartridge.go]: Corrected.

File: cli/replicaset/cmd/roles.go (hunk 19 above)
Comment (themilchenko) [reply cli/replicaset/cmd/roles.go]: Edited.

File: cli/replicaset/cmd/roles.go (hunk 20 above)
Comment (themilchenko) [reply cli/replicaset/cmd/roles.go]: Fixed.

**Review by oleg-jukovec — APPROVED** [review]

**Review by DerekBum — APPROVED** [review]
